File: solutions/python/perfect-numbers/1/perfect_numbers.py
from typing import Optional


def aliquot_sum(n: int) -> int:
    r = 0
    x = 2
    while x * x <= n:
        if n % x == 0:
            r += x
            if x * x < n:
                r += n // x
        x += 1
    return r + 1


def classify(number: int) -> Optional[str]:
    """A perfect number equals the sum of its positive divisors.

    :param number: int a positive integer
    :return: str the classification of the input integer
    """

    if number < 1:
        raise ValueError("Classification is only possible for positive integers.")

    if number == 1:
        return "deficient"

    asum = aliquot_sum(number)
    if asum < number:
        return "deficient"
    elif asum == number:
        return "perfect"
    else:
        return "abundant"

File: solutions/go/perfect-numbers/1/perfect_numbers.go
package perfect

import "errors"

type Classification string

var (
	ClassificationAbundant  Classification = "ClassificationAbundant"
	ClassificationDeficient Classification = "ClassificationDeficient"
	ClassificationPerfect   Classification = "ClassificationPerfect"

	ErrOnlyPositive = errors.New("ErrOnlyPositive")
)

func aliquot_sum(n uint) uint {
	var r uint
	var x uint = 2
	for x*x <= n {
		if n%x == 0 {
			r += x
			if x*x < n {
				r += n / x
			}
		}
		x += 1
	}
	return r + 1
}

func Classify(n int64) (Classification, error) {
	if n < 1 {
		return "", ErrOnlyPositive
	}
	if n == 1 {
		return ClassificationDeficient, nil
	}
	asum := aliquot_sum(uint(n))
	if asum < uint(n) {
		return ClassificationDeficient, nil
	}
	if asum == uint(n) {
		return ClassificationPerfect, nil
	}
	return ClassificationAbundant, nil
}
